net: dhcpv6: Add Zephyr DHCPv6 client

rlubos opened the pull request:
Add a DHCPv6 client implementation for Zephyr ([RFC 8415](https://www.rfc-editor.org/rfc/rfc8415)).

The implementation allows to request IPv6 address and/or prefix from the
DHCPv6 server, and for now supports only the mandatory set of DHCPv6
options needed to achieve this. Currently the implementation supports
the following scenarios:
 * Requesting new IPv6 address/prefix with Solicit/Request exchange
 * Refreshing existing leases with Confirm, Renew or Rebind (depending
   on the context).

For now, no Information Request (the case where neither IPv6 address or
prefix are requested) is supported. No support for Reconfigure was added
either, as this is optional (the client manifests clearly to the server
that it does not support Reconfigure). Support for these can be added
later if needed.

~~I'm still working on test suite for DHCPv6, so creating as a draft for now for early review. 
I'll update the PR once tests are ready.~~

rlubos committed:
net: dhcpv6: Add Zephyr DHCPv6 client

Add a DHCPv6 client implementation for Zephyr (RFC 8415).

The implementation allows to request IPv6 address and/or prefix from the
DHCPv6 server, and for now supports only the mandatory set of DHCPv6
options needed to achieve this. Currently the implementation supports
the following scenarios:
 * Requesting new IPv6 address/prefix with Solicit/Request exchange
 * Refreshing existing leases with Confirm, Renew or Rebind (depending
   on the context).

For now, no Information Request (the case where neither IPv6 address or
prefix are requested) is supported. No support for Reconfigure was added
either, as this is optional (the client manifests clearly to the server
that it does not support Reconfigure). Support for these can be added
later if needed.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: dhcpv6: Introduce net events for DHCPv6

Add net events for DHCPv6, in similar fashion as it's done for
DHCPv4.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: config: Add DHCPv6 support

Add DHCPv6 support to the net config library. In case DHCPv6 is enabled,
net config will attempt to acquire IPv6 address and/or prefix when used.
The user can select with Kconfig whether to request address or prefix
(or both).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: shell: Add DHCPv6 support

Print DHCPv6 status in net shell (with "net iface" command).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar reviewed (commented):
Some initial comments, did not manage to review all of it now. Overall looks great.

rlubos commented on include/zephyr/net/dhcpv6.h:
```diff
@@ -0,0 +1,112 @@
+struct net_dhcpv6_duid_storage {
+	uint8_t length;
+	struct net_dhcpv6_duid_raw duid;
+};
```
True, I've hidden the internal structs/defines.

rlubos commented on include/zephyr/net/net_if.h:
```diff
@@ -275,6 +278,73 @@ struct net_if_ipv6 {
 };
 
+#if defined(CONFIG_NET_DHCPV6) && defined(CONFIG_NET_NATIVE_IPV6)
+struct net_if_dhcpv6 {
```
Good point, there were some holes indeed, I've rearranged the fields and got down to 1 byte of padding.

rlubos commented:
I've added a commit with DHCPv6 test suite, hence moving the PR out of draft stage.

ssharks commented on subsys/net/lib/config/init.c:
```diff
@@ -249,6 +268,24 @@ static void ipv6_event_handler(struct net_mgmt_event_callback *cb,
 			 net_addr_ntop(AF_INET6, &laddr, hr_addr, NET_IPV6_ADDR_LEN));
+
+		if (ifaddr->addr_type == NET_ADDR_DHCP) {
+			char remaining_str[sizeof("01234567890")];
```
Why the sizeof this string?

rlubos replied:
Oh, it was just a copy paste of the code from [net shell](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_shell.c#L3468) where similar approach was taken, I didn't give too much though about it.

Indeed it might not be the best option, not very intuitive, I can rework this part to use "lifetime" string and it's size as a default, the buffer should be large enough to hold lifetime values (represented in seconds) in years.

rlubos replied:
I've pushed an update.

ssharks reviewed (commented):
Look like a valuable addition and a lot of work, one remake I found.

jukkar commented on subsys/net/ip/dhcpv6.c:
```diff
@@ -0,0 +1,2205 @@
+	if (mgmt_event == NET_EVENT_IF_DOWN) {
+		NET_DBG("Interface %p going down", iface);
+		dhcpv6_set_timeout(iface, UINT64_MAX);
+		dhcpv6_reschedule();
```
As both branches call reschedule, we could move the function call outside of the if statements.

jukkar commented on subsys/net/ip/dhcpv6.c:
```diff
@@ -0,0 +1,2205 @@
+
+	switch (iface->config.dhcpv6.state) {
+	case NET_DHCPV6_DISABLED:
+		NET_ERR("DHCPv6 already disabled on iface %p", iface);
```
Hmm, is this really an error? From application side it is, but from dhcpv6 client point of view this is just a no-op. Would warn or even info be enough here?

rlubos replied:
True, downgraded to NET_INFO level.

jukkar commented on subsys/net/ip/dhcpv6.c:
```diff
@@ -0,0 +1,2205 @@
+
+	sys_slist_init(&dhcpv6_ifaces);
+
+	/* All_DHCP_Relay_Agents_and_Servers (ff02::1:2) */
+	net_ipv6_addr_create(&all_dhcpv6_ra_and_servers,
+			     0xff02, 0, 0, 0, 0, 0, 0x01, 0x02);
+
+	/* All_DHCP_Servers (ff05::1:3) */
+	net_ipv6_addr_create(&all_dhcpv6_servers,
+			     0xff05, 0, 0, 0, 0, 0, 0x01, 0x03);
```
No need for these if we set the addresses at the start of the file.

jukkar reviewed (dismissed):
Overall look very good, very minor nits found during review.

rlubos commented on subsys/net/ip/dhcpv6.c:
```diff
@@ -0,0 +1,2205 @@
+static K_MUTEX_DEFINE(lock);
+
+static struct in6_addr all_dhcpv6_ra_and_servers;
+static struct in6_addr all_dhcpv6_servers;
```
Good point, it actually allowed me to spot that `all_dhcpv6_servers` is not needed at all in our case (used only by Relay Agents) so removed it.

rlubos commented:
Thanks @jukkar, I've addressed your feedback.

jukkar commented:
Just trying to find reviewers for this one, if possible please take a look. @pdgendt @cfriedt

pdgendt commented on subsys/net/ip/dhcpv6.c:
```diff
@@ -0,0 +1,2206 @@
+static const struct in6_addr all_dhcpv6_ra_and_servers = { { { 0xff, 0x02, 0, 0, 0, 0, 0, 0,
+							       0, 0, 0, 0, 0, 0x01, 0, 0x02 } } };
+
+static sys_slist_t dhcpv6_ifaces;
```
Use init macro:

```suggestion
static sys_slist_t dhcpv6_ifaces = SYS_SLIST_STATIC_INIT(&dhcpv6_ifaces);
```

rlubos replied:
Thanks, updated

cfriedt commented on tests/net/dhcpv6/src/main.c:
```diff
@@ -0,0 +1,1267 @@
+/*
```
Thanks for writing tests :-)

Just a suggestion - not blocking by any means - but it might be helpful to separate different aspects of testing into different files. Keep the `main.c` for general setup-before-after, teardown and state.

It might be easier to know where to put additional tests in the future.

cfriedt reviewed (dismissed):
Just a minor housekeeping suggestion for the testsuite (btw, thank you for adding the testsuite). Not blocking.

Otherwise LGTM 👍

carlescufi commented on include/zephyr/net/dhcpv6.h:
```diff
@@ -0,0 +1,116 @@
+	NET_DHCPV6_REBINDING,
+	NET_DHCPV6_INFO_REQUESTING,
+	NET_DHCPV6_BOUND,
+} __packed;
```
I don´t think so. There's a separate option for compilers that decides on the size of an enum type I believe.

rlubos replied:
At least for [GCC](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/Common-Type-Attributes.html), it makes the underlying type the smallest possible, regardless of the compiler option:
> When attached to an enum definition, the packed attribute indicates that the smallest integral type should be used. Specifying the -fshort-enums flag on the command line is equivalent to specifying the packed attribute on all enum definitions.

rlubos committed:
tests: net: dhcpv6: Add tests

Add tests for the DHCPv6 library. The tests verify that:
* message format is valid for the outgoing messages
* incoming messages are only processed in respective states
* processing through standard DHCPv6 transactions is correct
  (Solicit/Confirm/Renew/Rebind)

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>
